promql/prettier/node.go

```diff
@@ -0,0 +1,163 @@
+package prettier
+
+import (
+	"github.com/prometheus/prometheus/promql/parser"
+	"reflect"
+)
+
+type nodeInfo struct {
+	head parser.Expr
+	// Node details.
+	columnLimit int
+	history     []reflect.Type
+	item        parser.Item
+}
+
+func newNodeInfo(head parser.Expr, columnLimit int, item parser.Item) *nodeInfo {
+	return &nodeInfo{
+		head:        head,
+		columnLimit: columnLimit,
+		item:        item,
+	}
+}
+
+func (p *nodeInfo) violatesColumnLimit() bool {
+	return len(p.head.String()) > p.columnLimit
+}
+
+// getNode returns the node corresponding to the given position range in the AST.
+func (p *nodeInfo) getNode(root parser.Expr, item parser.Item) parser.Expr {
+	history, node, _ := p.nodeHistory(root, item.PositionRange(), []reflect.Type{})
+	p.history = history
+	return node
+}
+
+func (p *nodeInfo) baseIndent(item parser.Item) int {
+	history, _, _ := p.nodeHistory(p.head, item.PositionRange(), []reflect.Type{})
+	history = reduceContinuous(history, "*parser.BinaryExpr")
+	return len(history)
+}
+
+// contains verifies whether the current node contains a particular entity.
+func (p *nodeInfo) contains(element string) bool {
+	switch element {
+	case "grouping-modifier":
+		switch n := p.head.(type) {
+		case *parser.BinaryExpr:
+			return len(n.VectorMatching.MatchingLabels) > 0 || n.ReturnBool
+		}
+	case "scalars":
+		switch n := p.head.(type) {
+		case *parser.BinaryExpr:
+			if n.LHS.Type() == parser.ValueTypeScalar || n.RHS.Type() == parser.ValueTypeScalar {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+// parentNode returns the parent node of the given node/item position range.
+func (p *nodeInfo) parentNode(head parser.Expr, rnge parser.PositionRange) reflect.Type {
+	ancestors, _, found := p.nodeHistory(head, rnge, []reflect.Type{})
+	if !found {
+		return nil
+	}
+	if len(ancestors) < 2 {
+		return nil
+	}
+	return ancestors[len(ancestors)-2]
+}
+
+// nodeHistory returns the ancestors of the node the item position range is passed of,
+// along with the node in which the item is present. This is done with the help of AST.
+// posRange can also be called as itemPosRange since carries the position range of the lexical item.
+func (p *nodeInfo) nodeHistory(head parser.Expr, posRange parser.PositionRange, stack []reflect.Type) ([]reflect.Type, parser.Expr, bool) {
+	var nodeMatch bool
+	switch n := head.(type) {
+	case *parser.ParenExpr:
+		nodeMatch = true
+		stack = append(stack, reflect.TypeOf(n))
+		if n.PositionRange().Start <= posRange.Start && n.PositionRange().End >= posRange.End {
```

Comment:
There's a lot of code duplication here. You should use the `parser.Children` function after the switch instead of reimplementing it.

Reply:
Sure. I will take a look at this later on when the implementation of all types are done. After that, efficiency would be a key requirement. However, do note the importance of stack here. I hope, similar things are these in Children (that solve the functional purpose here) without much re-write.

Reply:
Please do these things early, since such monster functions are error prone and hard to review.

Reply:
The `parser.Children` returns the list of children of a node. But the requirement here is the list of parents or ancestors in order. This is because we need to find the depth of a node in the AST, for which parents list are required.

Reply:
Having a generic way of finding child nodes still can drastically simplify the function, large parts of it can be replaced by a

```go
for _, node := range parser.Children(n) {
...
}
```

after the switch case.

Reply:
Also, there is yet another thing to consider. Finding children nodes from a `node` is simple since it involves straight forward recursion. But, finding parents/children from `item` is not that straight forward. We need to compare the posRange and I think that's where `parser.Children()` will not be healthy. Any views?

Reply:
Do the comparison inside the for loop and `break` or recurse once you found the right child. Similar to 

https://github.com/prometheus-community/promql-langserver/blob/76204d7722d06ec84ea5851c4d42aac840abb9cd/langserver/cache/ast.go#L24-L47

Reply:
I don't think that would be neat. If you see my very first commits, I used to do that way, using a loop and going inside the children (manually) and at the same time inserting the node into a stack so as to get the list of ancestors. But, the situation got worse when Binary expressions came into the scenario since then conditional breaks were used and that looked untidy. Further to add on, in functional calls, we have more than 2 exprs, that means more added difficulties. 

The issue remains that in case of `(label_join(up{job="api-server",src1="a",src2="b",src3="c"}, "foo", ",", "src1", "src2", "src3")`, when you encounter `,`, how do you come to know whether it is in the current node, or the expr around it? The answer is very simple and clean in a **recursive approach** since we just need to call again recursively and see if it was there or not. So in this case, if the surrounding nodes report `false` to the existence of `,`, we come to know that it is from the current node (i.e., `parser.Call`). This was one of the reasons why I switched from for loop to a better recursive way.

Also, the current recursive approach is very straight forward. It just checks if an item is in the posRange of that node or not, using the item's posRang and then makes the decision of whether or not to add to the stackHistory.

Reply:
I'm not against using recursion.

The problem is, that the implementation of finding child nodes already exists in the parser package and is duplicated here. Which nodes you recurse into shouldn't be determined inside the `switch`, but using the `Children` function.

Reply:
More concretely:

```go
switch n := head.(type) {
    // Type specific handling
    ...
}

for _, child := range parser.Children(head) {
    if n.PositionRange().Start <= posRange.Start && n.PositionRange().End >= posRange.End {
        return p.nodeHistory(child, posRange, stack)
    }
}
```

Reply:
I still wonder how effective would that be in case of operators (and with grouping) in Binary expression or in case of commas in a function call.

Reply:
What do you mean by "effective"?

Reply:
I have implemented the Children function and now that greatly simplifies the code in terms of understanding. Thank you for your guidance.